### PR TITLE
V2.0

### PR DESCRIPTION
**Pull request by destrys**
1. Trezor firmware 1.6.2 compatible
2. Ledger support
3. Trezor Connect v5

### Trezor Compatibility
With trezor/trezor-mcu@1f470cf Trezors now sign with the same algorithm as other ethereum wallets, greatly simplifying our contract's ability to support multiple wallet vendors.

The contract signature verification has been updated to be compatible with Trezor firmware 1.6.2. **This breaks compatibility with Trezors running firmware 1.6.1 and below**

### Ledger Support
Our contract now supports the Ledger Nano S!

Currently, there are limitations on the BIP32 paths available to ledgers, and warnings are included during contract creation. 

While signing, the expected message displayed on the Ledger is provided.

The text throughout the contract and dapp has been updated now that it is no-longer Trezor-only. **The contract is now MultiSig2of3 instead of TrezorMultiSig2of3.**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -71,6 +72,9 @@ testrpc:
 server:
```
Comment (dhruvbansal) [inline Makefile]: We should remove this target and rename the `https-server` target to `server`, no?

File: README.md [hunk 2]
```diff
@@ -1,6 +1,8 @@
 Ethereum Multisig
```
Comment (dhruvbansal) [inline README.md]: Let's fix the language in this README before we merge.

File: contracts/MultiSig2of3.sol [hunk 3]
```diff
@@ -34,68 +34,96 @@ contract TrezorMultiSig2of3 {
...
     return message;
   }
-  
+
   // Send the given amount of ETH to the given destination using
   // the two triplets (v1, r1, s1) and (v2, r2, s2) as signatures.
   // s1 and s2 should be 0x00 or 0x01 corresponding to 0x1b and 0x1c respectively.
-  function spend(address destination, uint256 value, uint8 v1, bytes32 r1, bytes32 s1, uint8 v2, bytes32 r2, bytes32 s2) public {
+  function spend(address destination,
+		 uint256 value,
```
Comment (dhruvbansal) [inline contracts/MultiSig2of3.sol]: Let's put the each (v, r, s) triple on 1-line.

File: contracts/MultiSig2of3.sol [hunk 4]
```diff
@@ -34,68 +34,96 @@ contract TrezorMultiSig2of3 {
...
+		 bytes32 s2
+		 ) public {
     // This require is handled by generateMessageToSign()
     // require(destination != address(this));
-    require(this.balance >= value);
-    require(_validSignature(destination, value, v1, r1, s1, v2, r2, s2));
+    require(address(this).balance >= value);
+    require(_validSignature(destination,
+			    value,
+			    v1,
```
Comment (dhruvbansal) [inline contracts/MultiSig2of3.sol]: Let's put each (v, r, s) triple on one line.

File: contracts/MultiSig2of3.sol [hunk 5]
```diff
@@ -34,68 +34,96 @@ contract TrezorMultiSig2of3 {
...
-    Spent(destination, value);
+    emit Spent(destination, value);
   }
 
   // Confirm that the two signature triplets (v1, r1, s1) and (v2, r2, s2)
   // both authorize a spend of this contract's funds to the given
   // destination address.
-  function _validSignature(address destination, uint256 value, uint8 v1, bytes32 r1, bytes32 s1, uint8 v2, bytes32 r2, bytes32 s2) private constant returns (bool) {
+  function _validSignature(address destination,
+			   uint256 value,
```
Comment (dhruvbansal) [inline contracts/MultiSig2of3.sol]: Let's put each (v, r, s) triple on one line.

File: Makefile (hunk 1 above)
Comment (destrys) [reply Makefile]: done.

File: README.md (hunk 2 above)
Comment (destrys) [reply README.md]: updated.

**Comment by destrys** [comment]
Realized I forgot about the ledger scripts. I've updated those and put code blocks in the README. 

Also bumped the mythril version and added a solium command

**Comment by destrys** [comment]
solium now passes cleanly. Contracts updated to abide by the solidity style guide

**Comment by destrys** [comment]
went a little crazy, but now runs through all solium tests cleanly

**Comment by destrys** [comment]
Note: a new breaking change is the Funded event signature has changed to Funded(uint newBalance)

**Comment by howech** [comment]
I read the contract, build the dapp and rand through some scenarios. Everything seems to be working. Left some minor feedback with Destry.

**Comment by destrys** [comment]
Three new features/fixes on their way:
1. Spend broadcasted dialog should actually poll for the transaction and either return success or failure.
2. Changing require() error messages to codes (e.g. "3") with descriptions in the header comment
3. Clicking any of the navbar links from create or spend pages will open a new window so that you will have to purposefully close those windows. (To make it even harder to lose the bip32 paths you entered)

**Pull request merged**